### PR TITLE
Strict mode improvements

### PR DESCRIPTION
**Pull request by karate**
Some improvements regarding the strict-mode option

### CODE REVIEW

**Comment by PatricNox** [comment]
Hello!

Please ensure the features are functional before opening a PR.

`Notice: Trying to get property 'num_rows' of non-object in /Users/lia/playground/search/dbSearcher/dbSearcher.php on line 79`

File: index.php [hunk 1]
```diff
@@ -13,10 +13,10 @@
 	// Check if there are inputs
 	if (isset($_POST['dbs-database'], $_POST['dbs-search']))
 	{
-	    $search = $_POST['dbs-search'];
+    $search = $_POST['dbs-search'];
```
Comment (PatricNox) [inline index.php]: This indentation is wrong.

File: index.php [hunk 2]
```diff
@@ -13,10 +13,10 @@
...
 		$_SESSION['query'] = $_POST['dbs-search'];
 		$_SESSION['database'] = $database = $_POST['dbs-database'];
-		$strict = ($_POST['dbs-strict']) ? TRUE : FALSE;
+		$strict = (isset($_POST['dbs-strict'])) ? TRUE : FALSE;
```
Comment (PatricNox) [inline index.php]: Better to use !empty here.

File: dbSearcher/template.php [hunk 3]
```diff
@@ -9,7 +9,8 @@
 		<div class="clearfix">
 		<div class="strict float-r">
       <label for="dbs-strict">Find exact match</label>
-			<input type="checkbox" name="dbs-strict" id="dbs-strict" value="TRUE" />	
+      <?php $checkbox_attributes = $strict? 'checked': '' ?>
+      <input type="checkbox" name="dbs-strict" id="dbs-strict" value="TRUE" <?php echo $checkbox_attributes; ?>/>
```
Comment (PatricNox) [inline dbSearcher/template.php:13]: Nice thought! but this should be in a session just like query and database. See index.php

File: dbSearcher/dbSearcher.php [hunk 4]
```diff
@@ -76,7 +76,7 @@ function dbs_search($search, $database, $strict = FALSE)
 					$sql_search .= implode(" OR ", $sql_search_fields);
 					$rs3 = $mysqli->query($sql_search);
 				
-					if ($rs3->num_rows > 0) {
+					if ($rs3 && $rs3->num_rows > 0) {
```
Comment (PatricNox) [inline dbSearcher/dbSearcher.php:79]: This is a hackfix. 

Underlying issue lies in this scope:

```
					if ($rs2->num_rows > 0) {
						while ($r2 = $rs2->fetch_array()){
							$column = $r2[0];

							if ($strict)
								$sql_search_fields[] = $column." = ".$search;
							else 
								$sql_search_fields[] = $column." LIKE('%".$search."%')";
						}
						
						$rs2->close();
					}
```

**Review by PatricNox — CHANGES_REQUESTED** [review]
Followed by the error above, there's a bit of changes needed in your PR.

**Comment by karate** [comment]
> Please ensure the features are functional before opening a PR.
> 
> `Notice: Trying to get property 'num_rows' of non-object in /Users/lia/playground/search/dbSearcher/dbSearcher.php on line 79`

I cannot reproduce this error. I think commit bdb72a4  fixes it.

File: index.php (hunk 2 above)
Comment (karate) [reply index.php]: i think that `isset` is better that `!empty` when checking if an array key exists. But I can change it to `!empty` if you insist. Or, even better, use [array_key_exists](https://www.php.net/manual/en/function.array-key-exists.php)

File: dbSearcher/template.php (hunk 3 above)
Comment (karate) [reply dbSearcher/template.php:13]: Ok. I can do this.

**Comment by PatricNox** [comment]
Query with the strict mode enabled and you'll get the message. 
I reproduced through your forked branch.

The issue also resides within the mentioned while loop.

**Comment by karate** [comment]
I still cannot reproduce this.
I tried with both states of strict mode. With queries that return results, and queries that don't.
I'm using php 7.0.

**Comment by PatricNox** [comment]
Hmmm interesting. Which database server are you using?

The issue is anyways the query who needs to be wrapped with apostrophes.

`$sql_search_fields[] = $column." = ".$search`

should be

`$sql_search_fields[] = $column." = 'search'"`

otherwise the query is as following and therefore failing:

` SELECT * FROM table WHERE column = searchkey AND column2 = searchkey AND ...`

File: index.php (hunk 2 above)
Comment (PatricNox) [reply index.php]: It can exist but still being false. As we want to be strict, it's better to use empty function here.

**Comment by karate** [comment]
Ok, I managed to get
`Notice: Trying to get property 'num_rows' of non-object in /usr/src/DatabaseSearcher/dbSearcher/dbSearcher.php on line 54`
when hitting Search with an empty database value.

File: dbSearcher/dbSearcher.php (hunk 4 above)
Comment (karate) [reply dbSearcher/dbSearcher.php:79]: I think this is outside the scope of the current PR. The problem you mentioned in https://github.com/PatricNox/DatabaseSearcher/pull/14#issuecomment-542860139 is a more generic problem on variable escaping.

I'm preparing a new PR which rewrites the `dbs_search` function and adds quotes in variables and some more descriptive variable names, if you're interested.

**Comment by PatricNox** [comment]
Closing as this will be included in https://github.com/PatricNox/DatabaseSearcher/pull/17

**Pull request closed without merging**